For review

jhford opened the pull request:
This is the first pull request for the new Artifact API.  I have decided to split the S3 specific pieces into their own library.  This is code that besides being used by the Queue is completely unrelated to it.  It is possible that someone else might wish to use it.  It is also a lot easier to develop as an independent library.

The general architecture is that the Queue will maintain an instance of the `Controller` class from this library.  It will call into that library to run certain operations locally on the queue, and others to generate objects in what I've called "Interchange Format Objects", or `{url, headers, method}` in JS terms.

The `Client` class will be used to implement the JS library which will eventually be used in Docker Worker and published to NPM for other users.  The `Runner` class is used by both `Client` and `Controller` to perform the underlying HTTP/HTTPS operations.

I've chosen to use the Joi library to do object validation for all functions.  I've found this to be extremely helpful in diagnosing weirdness.  It has caught a *ton* of bugs while doing development.

There is a unit test suite through `npm test` which runs tests against actual mock HTTP servers.  I've implemented a basic mocking of the S3 API where it's needed.  This is not an attempt to become a general purpose S3 Mock.

There's still a few things which I need to implement.  Please take this into account when doing the review.  If anyone wishes to help out on these features, I'm happy to work with them.

* Support for `100 CONTINUE` on upload operations.  This allows us to ensure the upload will succeed before sending the whole body.  This is an optimization.  This is irrelevant to the server-side portions of the library
* Concurrency for `Runner`.  When running a list of requests, the `Runner` should be able to run requests in parallel, but only up to a maximum number of concurrent requests.  It need to ensure that the results map back to the order of requests.
* <strike>Retry for `Runner.prototype.run()`.  I haven't implemented it yet.  One thing to note is that I support passing a closure as the parameter for the body so that even streaming requets can be retried.  Retry logic should not be allowed when passing a stream directly.</strike>7bd440e
* Download support: this is primarily an issue of implementing 300 redirect handling
* Tests regarding hangups and aborts which ensure that the appropriate error is shown for each permutation of
  * streaming request vs buffer request
  * streaming response vs buffer response
  * client hangup vs server hangup
* Setup of the linter
* fix up example.js

In this review, I'm looking for feedback on the following things:

* Whether the public API presented by the `Controller` class has any papercuts
* Whether there are test missing which are not on the above list of todo items
* Whether there are bugs or other technical issues with the codebase
* Whether there are error cases which have not been explore
* Whether the S3 Permissions and security model is being implemented correctly
* Whether the S3 Rest API is being used correctly
* Dead code
* Feedback is based on technical aspects

In this review, I am not looking for:

* Feedback on personal preference about coding style where there is not a clear technical backing
* Requests to rewrite the entire library, or comments that amount to a request to rewrite
* Requests to completely re-architect the library
* Requests to use different/fewer/more libraries where there isn't a technical reason
* Things that our lint configs will catch -- I'll fix those up when I get the linter installed

I would highly recommend not looking at intermediate commits unless you're really interested.  You should be able to run 'npm test' without any credentials.  There's a file `src/example.js` which includes an example implementation.  This is not 100% functional right now, but it does show a rough overview of how to implement the library.  My intention is to include that in the README.md file eventually through a prepublish flag

NOTE: I've asked a bunch of you to review the library.  The primary reviewer for this code is @jonasfj, but I would appreciate if others could take a look as well.  If you're not @jonasfj, please feel free to ignore :)

Thanks!

jhford committed:
deps

jhford committed:
Initial implementation

jhford committed:
First XML tests

jhford committed:
XML Parsing Unit Tests complete

jhford committed:
XML Tests Complete

jhford committed:
Unit tests for methods which call S3 directly

jhford committed:
Tests for generating requests to run remotely

jhford committed:
Add status code to the runner response

jhford committed:
Tests for XML.  Forgot to git add them :/

jhford committed:
Initial implementation of the run function

jhford committed:
Implement the run() function

jhford committed:
travis

jhford committed:
Move run() into a new class, Runner

jhford committed:
Ensure SHA256 and Size validity

jhford committed:
pass objects into functions instead of raw args

jhford committed:
Return the etag from a multipart upload

jhford committed:
Obtain the ETag from the complete multipart artifact for storing on the server

jhford committed:
Tests for DigestStream

jhford committed:
Rename S3 class to Controller

jhford committed:
README.md

jhford committed:
only export things we really want to export

jhford committed:
use g++ 4.8 so libxmljs might work

jhford committed:
Working on the run() method

jhford committed:
Clean up request logging

jhford committed:
Enforce interchange format and clean up runner output a little

jhford committed:
Improve the Runner.run interface

jhford committed:
support streaming response bodies

jhford committed:
Use opts for the Controller

jhford committed:
Work to allow for exponential backoff

jhford committed:
WIP Done for the evening

TODO:

* finish writing the single part and multipart upload client
* need to add assertions to all methods to ensure that options
  are passed in correctly
* maybe write some unit tests to test for server hangups and similar
* write unit tests which throw errors

jhford committed:
multipart preperation working

jhford committed:
multipart uploads

jhford committed:
Add assertions about multipart uploads -- basically <10000 parts and >5mb parts

jhford committed:
Implementing an example of the library

jhford committed:
Fix ETag parsing for multipart uploads

jhford committed:
Add support for permissions setting and tagging of resources

jhford committed:
Example working with single part uploads

jhford committed:
Last thing with multi-part uploads is signature validation errors on copmlete message

jhford committed:
Ensure that the content-sha256 matches the value we expect

jhford committed:
Refactor the .run() method to be cleaner

jhford committed:
Fix very strange issue with signature mismatch on complete upload message

jhford committed:
get rid of some old code that was commented out

jhford committed:
Assert more about the input to functions and use a local http server for tests

jhford committed:
delete some unneeded logging messages

jhford committed:
Handle case in test where last part is a full part

jhford committed:
Test multiple part sizes

jhford committed:
Cleanup

jhford committed:
Rewrote the Controller tests using a significantly improved mock.

This mock understands a lot more about how S3 actually works

jhford committed:
Use memory more efficiently

jhford committed:
Better tests, setup and teardown

jhford committed:
Stop using httpbin altogether

jhford committed:
Make sure that the errors in Controller are properly checked

jhford committed:
I forgot to add schemas

jhford committed:
Create bigfile if not there

jhford committed:
Export the client

jhford commented:
Note the travis issues are because I apparently need to use a hostname other than localhost there.  Anyone know what travis calls localhost? 
```
Uncaught Error: getaddrinfo ENOTFOUND test-bucket.localhost test-bucket.localhost:8080
```

jonasfj commented on README.md:
```diff
@@ -0,0 +1,138 @@
+  region specified by `region`.  The `runner` parameter is a function with the
+  signature `(request, body)` which must take a standard interchange format
+  object and returns an object like `Runner.prototype.run`
+* `Controller.prototype.initiateMultipartUpload(opts)`: initiate a multipart
```
aren't you missing md5?

jonasfj commented on README.md:
```diff
@@ -0,0 +1,138 @@
+* `new Controller(region, runner)`: Create a `Controller` instance for the
+  region specified by `region`.  The `runner` parameter is a function with the
+  signature `(request, body)` which must take a standard interchange format
+  object and returns an object like `Runner.prototype.run`
```
What should the object have?

jhford replied:
I suspect that I need to update this comment, i changed how the Runner is passed in.

jonasfj commented on README.md:
```diff
@@ -0,0 +1,138 @@
+  store the file in as well as `sha256` which is the SHA256 digest of the full
+  file and `size` which is the number of bytes in the full file.  This method
+  returns the string of of the `uploadId` to use for following requests.
+* `Controller.prototype.generateMultipartRequest(opts)`: generate a list of
```
again missing md5, right?

jonasfj commented on README.md:
```diff
@@ -0,0 +1,138 @@
+  headers: {
+    aresponse: "header"
+  },
+  statusCode: 200,
```
Wouldn't it be sufficient and complete to just get `{statusCode, headers, body}`, and specifying that body must be parsed as string.

jhford replied:
Except that for downloading (which will be implemented in this library shortly) will require streaming bodies.  It originally did what you said, but I realised that I needed download support.

jonasfj commented on README.md:
```diff
@@ -0,0 +1,138 @@
+
+Note that the request body itself is not specified anywhere in this format.
+
+The `Runner` class has the following methods:
```
You pass this as a function... could we calll it a `request` function...
Say a function `request({method, url, body}) => {statusCode, headers, body}`.
Where result body is a string, and input body is `Buffer` or `Readable` stream.

jhford replied:
Is there a technical reason to rename the function?

jonasfj commented on README.md:
```diff
@@ -0,0 +1,138 @@
+which must be run from the server.  The methods which must be run from the
+server are the initate, complete and abort methods of the multipart uploads.
+
+The `Controller` class has the following methods:
```
The docs here are lovely, but I would appreciate a more concise structure like:

```md
Controller#initiateMultipartUpload({uploadId, sha256, size, parts: [{sha256, size}, ...]}) -> [{method, url, headers}, ...]
```

I don't like having to read text to find the what the signature is.
Structure multiline thing like this works too:
```md
Controller#initiateMultipartUpload({
  uploadId,      // 8 word comment...
  sha256,        // this sha256 of everything in hex
  size,
  parts: [{sha256, size}, ...]}
) -> [
  {method, url, headers},
  ...
]
```
Or something like that... it's hard to get an overview by reading text.

jhford replied:
OK, but this is just newlines and a couple of clarifying comments.  I'm happy to restate the method signatures at the end of the doc if you'd like though.

jonasfj commented on src/controller.js:
```diff
@@ -0,0 +1,601 @@
+
+  // Return a list of 2-tuple's that are header key and value pairings for the
+  // headers to specify as the ACL for an object
+  __determinePermissionsHeaders(permissions) {
```
I think this is unwanted complexity... We only use bucket policies as is recommended by aws.
So we'll never want to set any permissions.

jhford replied:
we don't have to use it in the Queue, but it is there and tested.  This is a generic library and I have been using these permissions for integration tests that I'm working on.  If there's a specific test that you think is lacking, please let me know.

jonasfj replied:
Okay, that was my main concern about this... It seemed like a lot of code to drag into the queue... When we could do most of it with `aws-sdk`...

For the record: I'm super excited about the testability of the client implementation... This is something we've neglected a bit before :)

jhford replied:
> For the record: I'm super excited about the testability of the client implementation... This is something we've neglected a bit before :)

Me too!  And using the `Controller` class in the queue means that we make sure that our ability to sign and run S3 requests is always there, both by testing this library and by integreation testing the Queue :)

jonasfj commented on src/controller.js:
```diff
@@ -0,0 +1,601 @@
+      key: schemas.key.required(),
+      sha256: schemas.sha256.required(),
+      size: schemas.mpSize.required(),
+      permissions: schemas.permissions,
```
permissions is unwanted... In some future maybe we want storage class.

We definitely want to specify metadata...

jhford replied:
Ahh yes, I should add the ability to pass metadata into the methods.  Good catch.  I'll look into what's required to add storage class.

jonasfj commented on src/controller.js:
```diff
@@ -0,0 +1,601 @@
+      path: `/${key}?uploads=`,
+      headers: {
+        'x-amz-meta-content-sha256': sha256,
+        'x-amz-meta-content-length': size,
```
content-length is not a meta-data property, adding it as metadata just creates additional confusion.
Also `contentType` is missing.

jhford replied:
except that we sort of have to do it.

Content-Length is part of V4 signing when specified, so we need to pass the actual content-length for each call to AWS.  For single part uploads, that's no problem because the content-length is the same as the artifact size.  For multipart uploads, however, we need to set the content-length header to the artifact size, but that'd cause the signing of the initiate multipart upload to fail because that request would have the wrong content-length.

S3 might calculate a valid Content-Length at the end of the multipart upload, but it is not a value directly passed through by the worker.  That's why we need to have `x-amz-meta-content-length`: it's the exact length the worker thought the file is.

I set this for single part uploads as well so that they have the same structure.

I'll add support for content-type.

jonasfj replied:
Ah, my thinking was that S3 would calculate it and that would be enough...

I think it'll be confusing to have two... Maybe we can just validate the content-length that s3 reports to the we got from the worker..

Hmm, just looked it up and would be an extra head request....

But when we think about it for a bit... Each of the part upload requests includes a size... And those sizes are signed so they must match... And we can't finish without all parts .. so content-length can't be sorter than what was declared in createArtifact...  Even with transport errors... It would have to be an S3, which I don't think we should harden against.

jonasfj replied:
An S3 bug.

jhford replied:
One of the main motivations behind this work is to make it easier to detect corruption.  I think anticipating S3 bugs is a completely reasonable thing to do.  Having this header doesn't really break anything, and at worst is a little redundant.  Are you aware of technical consequence of having an `x-amz-content-length` header?

jonasfj commented on src/controller.js:
```diff
@@ -0,0 +1,601 @@
+    let signedRequest = aws4.sign(unsignedRequest);
+
+    let response = await this.runner({
+      req: this.__serializeRequest(signedRequest)
```
there is no need for all this complication... you can just use aws-sdk:
http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#createMultipartUpload-property

jhford replied:
It's not complicated at all.  Once you've implemented the code to generate a multipart part upload request and the code to run said request, it's actually trivially easy to call all methods the same way.  It also ensures that all of the signing code that will eventually be used for the workers works 100%.

I also think that the `aws-sdk` library is painful to work with.  This adds another dependency for no benefit.

Which test case is not covered that you feel should be?

jonasfj commented on src/controller.js:
```diff
@@ -0,0 +1,601 @@
+        headers: {
+          'x-amz-content-sha256': part.sha256,
+          'content-length': part.size,
+        }
```
Missing `content-md5`

jhford replied:
I suspect that like `content-length` we won't be able to use `Content-MD5` for multipart uploads.  If that's the case, we won't get any of the automatic stuff about the header, so I don't know if it's worth it to include MD5 when we already have SHA256.

jonasfj replied:
You can for each part specify the md5

jonasfj replied:
I guess it doesn't add much over sha256...

jhford replied:
that's sort of my thinking.  The only reason I wanted MD5s is to have the Content-MD5 header offered to clients, in the hope that because it's an HTTP spec and not a vendored prefix, it might trick clients into automatically doing MD5 sums for those who don't chose to do `x-amz-content-sha256` validation.

jonasfj commented on src/controller.js:
```diff
@@ -0,0 +1,601 @@
+   *
+   * http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUTtagging.html
+   */
+  async __tagObject(opts) {
```
Okay, tagging support is sort of cool... But
A) Do we want it? Considering that it's an extra API call for every artifact. I suspect it might cool to have.
B) Why not do this using `aws-sdk` there is no need to be making XML our selves...

jhford replied:
Yes, we want it.  For single part uploads it's atomic and part of the original PUT, so it's zero overhead.

Multipart uploads are large files to begin with, so the cost of an API call vs. the size of the file makes it pay off.  It also means that we're doing a lot of API calls, 2 as setup and teardown and 1 for each part.  An extra tagging call is nothing.

My main motivation was so that we could break down our costs in the Cost Explorer, and big files (so multipart) are the most important.

There's no reason to use aws-sdk.  The XML needed is trivially simple to generate and there's full unit tests for it.

jonasfj replied:
You say XML is trivial but wait till garndt start adding special characters or something...
My argument isn't that this code is bad or doesn't have sufficient test. But that we shouldn't maintain code when we don't have to. Arguably aws-sdk is better maintained...

Note: writing fast-azure-storage was not a 100% good idea. It worked out because the original client was extremely bad... But we've had a LOT of minor bugs in it. If avoidable we should.

jhford replied:
I didn't say XML is trivial, rather that the *tiny* subset of XML that I need in this code is trivial.

The only time that we generate XML is:

1. generating tags
1. etag list

For the first, that's all data passed in from the Queue, so I'm not concerned.
For the second, that's data obtained from the client no matter what based on a list of strings.  We're getting opaque ETags from the client and sticking them in an `<etag>` value.

I think an extra validation of the ETag value is in order, but since the AWS docs specify the format of the multipart ETags, this is pretty trivial:

> Entity tag that identifies the newly created object's data. Objects with different object data will have different entity tags. The entity tag is an opaque string. The entity tag may or may not be an MD5 digest of the object data. If the entity tag is not an MD5 digest of the object data, it will contain one or more nonhexadecimal characters and/or will consist of less than 32 or more than 32 hexadecimal digits.

Azure's API had significantly more complex XML needs and is a significantly more complex API to implement.


What are your specific concerns?

jonasfj commented on src/controller.js:
```diff
@@ -0,0 +1,601 @@
+    let requestBody = this.__generateCompleteUploadBody(etags);
+    let requestBodySha256 = crypto.createHash('sha256').update(requestBody);
+
+    let signedRequest = aws4.sign({
```
Again why not use `aws-sdk` to make this request... No need to mess around with signature generation, or XM for that matter...

jhford replied:
There's no reason to use aws-sdk.  The signatures are tested and the XML generation and parsing is tested.  Is there a test case that you think should be added?

jonasfj commented on src/controller.js:
```diff
@@ -0,0 +1,601 @@
+   }));
+
+    let {bucket, key, uploadId} = opts;
+    let signedRequest = aws4.sign({
```
Again, let's just use `aws-sdk`, we only need to do signature generation when we want to return the signature.

jhford replied:
There's no reason to use aws-sdk.  We need to generate a signature for all requests.  This is doing it with the aws4 library instead of the aws-sdk library, but it's doing the exact same thing under the hood.

jonasfj commented on src/controller.js:
```diff
@@ -0,0 +1,601 @@
+        'content-length': size,
+        'x-amz-content-sha256': sha256,
+        'x-amz-meta-content-sha256': sha256,
+        'x-amz-meta-content-length': size,
```
Don't add content length as meta-data.... 

Also missing `content-md5`.

jhford replied:
As i mentioned above, there are technical reasons to add it as metadata as well.

jonasfj commented on src/controller.js:
```diff
@@ -0,0 +1,601 @@
+
+    // Set any tags.  For once, AWS Tags are atomic!
+    if (tags) {
+      unsignedRequest.headers['x-amz-tagging'] = qs.stringify(tags);
```
Okay, now you can convince me to tag things... This is awesome! :)

I'm totally willing to pay for the extra request required to tag multi-part uploads those are big. As long as we get tagging for free for single part uploads...

@jhford, Nice find! this is really awesome! :)


Note: now I'm tempted to limit `task.tags` to 10 tags, 2kb combined with same character encoding as S3 tags, just so we can tag all artifacts with `task.tags`... Granted that might be too strict for `task.tags`. But we should look into it.

jhford replied:
Sounds like a good idea! I'm happy to implement those checks.  We can always increase limits later on :)

jonasfj replied:
For the record we just started using task.tags for something else too... Actions.json notably... So there is pros/cons... Maybe we have a way to select tags we want reflect... And we set those tags in queue config... Maybe...

jonasfj commented on src/controller.js:
```diff
@@ -0,0 +1,601 @@
+ *
+ * http://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html
+ */
+function parseS3Response(body, noThrow = false) {
```
No need for any of this if we use `aws-sdk` for the requests we make directly.

jhford replied:
There's no need to use aws-sdk.  This is pretty simple code and it has tests.

jonasfj commented on src/digest-stream.js:
```diff
@@ -0,0 +1,28 @@
+import stream from 'stream';
```
Don't extend builtin streams... See: https://www.npmjs.com/package/readable-stream

jhford replied:
This class is not intended to be used outside of this library.  Given that, does adding a dependency on that package really make sense?

jonasfj commented on src/example.js:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    let requests;
+    if (opts.parts) {
```
Instead of making `parts` optional... Please make it `parts.length === 1`, so we have:
 1) Single part upload if `parts.length === 1`
 2) Multipart upload if `parts.length > 1`

jhford replied:
having parts for a single part upload means we're duplicating 100% of the information.  There's also a difference between a single part upload and a multipart upload with a single part.

jonasfj reviewed (changes requested):
To sum up:
 1) We should add `content-md5` too
 2) No `x-amz-meta-content-length` that's just duplication
 3) I love the tagging support
 4) The `Client` object abstraction looks interesting - I love the idea of a client library we can use in worker and queue-tests,
 5) Can we please use `aws-sdk` and avoid XML manipulation, etc... We get more features, less things to test and review.

I gave it a quick try to see how hard it would be to implement the `ServerAPI` in `example.js` using `aws-sdk` and `aws4` directly. Checkout:
https://gist.github.com/jonasfj/59f655509741a725626fc0e047191d85#file-example-js-L44-L165
It's still missing some details, forgot tagging at end of multi-part upload, and clearly I'm not importing aws4 or instantiating it.

Given how relatively easy that is, I don't think it's worth while using the `Controller` in the queue. The `aws-sdk` is much more battle tested than anything we write. That said it's truly awesome to have a mock setup like this that allows for worker tests without S3. This will be particularly powerful when we have to do retries and handling of streams correctly.

petemoore commented:
@jhford Looks like you've done a huge amount of awesome work! I wish I knew node better to be able to provide constructive feedback, but I'm looking forward to learning all about the library next week in Berlin, and starting on the go client-side implementation. :+1:

jhford commented:
> 
> 
> To sum up:
> 
>     We should add content-md5 too

I am happy to do so, but I wanted to discuss that.  Given that, we'd lose the nice properties of that header for anything done with multipart uploads, I think there's limited value in using Content-MD5.  We're already getting all the S3 side verifications by using x-amz-content-sha256.  Do you think we should still have an md5 header?

>     No x-amz-meta-content-length that's just duplication

It makes sense because of multipart.  The idea is to verify the content the worker said it had generated, so I want to pass that value from the client no matter way.  S3 might generate the real ContentLength header for multipart uploads incorrectly, and this will allow us to detect that case.

>     I love the tagging support

Great!

>     The Client object abstraction looks interesting - I love the idea of a client library we can use in worker and queue-tests,

Yep!  And it's really nice that the Queue and Worker would be using the exact same Signing and HTTP underpinnings for both sides, so we know they work.

>     Can we please use aws-sdk and avoid XML manipulation, etc... We get more features, less things to test and review.

I don't think there's any reason to use aws-sdk.  The features that you've asked for are pretty easy to add.  This is adding another library to do the work we basically already can.

The XML involved in the S3 api is trivially simple and I've written unit tests for it already.  If there are test cases that I've missed, I'm happy to add those.

> I gave it a quick try to see how hard it would be to implement the ServerAPI in example.js using aws-sdk and aws4 directly. Checkout:
> https://gist.github.com/jonasfj/59f655509741a725626fc0e047191d85#file-example-js-L44-L165
> It's still missing some details, forgot tagging at end of multi-part upload, and clearly I'm not importing aws4 or instantiating it.

Interesting, but that's Queue specific implementation and this is a general purpose library.

> Given how relatively easy that is, I don't think it's worth while using the Controller in the queue. The aws-sdk is much more battle tested than anything we write. That said it's truly awesome to have a mock setup like this that allows for worker tests without S3. This will be particularly powerful when we have to do retries and handling of streams correctly.

If you have specific technical deficiencies, please note them.

jonasfj commented:
> If you have specific technical deficiencies, please note them.

I don't !!!! :) Hehehe, the code does look solid and I am impressed by the number of tests.

I guess my concern is more statistical than technical. In the sense that the code is solid, but more lines of code increases risk of bugs... Hence, a statistical argument for not doing what aws-sdk can do. That's all...

Okay, night night ..

jonasfj commented:
@jhford, another feature request is support for `ContentEncoding`, I think we currently support setting it because it's not controlled by the signed URL.

docker-worker and other use it to gzip compress some artifacts. When creating a `blob` type artifact we probably have to specify the `ContentEncoding`. It's probably worthwhile to compress logs. Not sure if workers should automatically compress everything else with gzip too, but that should probably just be up to the worker..

jhford committed:
Switch readme to specifying the method signatures instead of a prose description

jhford committed:
clarify the comment about the Runner class

jhford committed:
Support storage class

jhford committed:
Validate that S3 tags are specified correctly

jhford committed:
Support arbitrary metadata

jhford committed:
Allow setting arbitrary metadata

jhford committed:
Allow setting content-type, content-encoding and content-disposition

jhford committed:
Unit tests to ensure that XML injection doesn't work

jhford commented:
So I think that I've addressed the majority of the review requests that weren't on the list of TODO items from opening the PR.

Specifically:

* I changed Readme.md to have just a list of method signatures
* I clarified the comment about the Runner class, which was awkwardly phrased
* I added support for setting a storage type
* I added support for setting extra arbitrary metadata and added support to check that *all* metadata combined fits the S3 restriction of 2048 Bytes of UTF-8 encoded data and no more than 128 bytes for the keys and 256 bytes for the values.
* I added verifications that the tags combined meet the criteria from S3 of no more than 2048 Unicode characters (not bytes!) and no more than 10 tags
* I added support for `Content-Type`, `Content-Encoding` and `Content-Disposition`
* I added unit tests to ensure that the XML is generated in a way that we aren't open to XML injection.

Regarding tagging.  Given that the taskcluster limit is 10, which is also the AWS limit and that AWS gives special properties in the cost explorer (which is important), I don't think that they're a good fit with the 10 tags and 2KB restriction.  Using metadata is another option, but it does limit us to 256 Bytes of data in the value.  Without some sort of metadata header wrapper, that'd be pretty awkward and I don't think something we'd want to do.  I think that we'll actually want to encode the following information:

* Name: `taskId/runId/name`
* Owner: `$provisoner/$worker-type`
* taskId
* runId
* name

as tags.  Either way, I've encoded the maximums for S3 into this library, we can add further restrictions in the queue pretty easily.

jhford committed:
fix broken package.json

jhford committed:
allow both vhost and path based addressing

This might make it easier to do unit testing on machines where we cannot
set up subdomains, e.g. unit tests with the mocked out S3

jhford committed:
Also check that the number of Part nodes is equal to the number of ETags passed into the generation

jhford committed:
Gitignore

jhford committed:
Include files explicitly

jhford committed:
Do automatic retries for 500 series errors in Runner.prototype.run().

This does not throw errors because that's established convention in this library

jhford commented:
@jonasfj I've implemented retrying logic per the recommendation of S3 in 7bd440e

jhford committed:
extra docs and remove the .only

jonasfj commented:
taskId is likely a bad tag, creates more noise that info...

Den 6. mar. 2017 8.54 AM skrev "John Ford" <notifications@github.com>:

> So I think that I've addressed the majority of the review requests that
> weren't on the list of TODO items from opening the PR.
>
> Specifically:
>
>    - I changed Readme.md to have just a list of method signatures
>    - I clarified the comment about the Runner class, which was awkwardly
>    phrased
>    - I added support for setting a storage type
>    - I added support for setting extra arbitrary metadata and added
>    support to check that *all* metadata combined fits the S3 restriction
>    of 2048 Bytes of UTF-8 encoded data and no more than 128 bytes for the keys
>    and 256 bytes for the values.
>    - I added verifications that the tags combined meet the criteria from
>    S3 of no more than 2048 Unicode characters (not bytes!) and no more than 10
>    tags
>    - I added support for Content-Type, Content-Encoding and
>    Content-Disposition
>    - I added unit tests to ensure that the XML is generated in a way that
>    we aren't open to XML injection.
>
> Regarding tagging. Given that the taskcluster limit is 10, which is also
> the AWS limit and that AWS gives special properties in the cost explorer
> (which is important), I don't think that they're a good fit with the 10
> tags and 2KB restriction. Using metadata is another option, but it does
> limit us to 256 Bytes of data in the value. Without some sort of metadata
> header wrapper, that'd be pretty awkward and I don't think something we'd
> want to do. I think that we'll actually want to encode the following
> information:
>
>    - Name: taskId/runId/name
>    - Owner: $provisoner/$worker-type
>    - taskId
>    - runId
>    - name
>
> as tags. Either way, I've encoded the maximums for S3 into this library,
> we can add further restrictions in the queue pretty easily.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/taskcluster/remote-signed-s3/pull/1#issuecomment-284458340>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJI5HkXP1G8JGMAGx7MADgzWtVK7JiRks5rjDotgaJpZM4MRmCO>
> .
>

jhford commented:
how we set the name/owner is something we can do in the Queue, i'll include that in the Queue PR

jonasfj commented on src/client.js:
```diff
@@ -0,0 +1,275 @@
+  }
+
+
+  async __prepareSinglepartUpload(opts) {
```
singlepart upload is just multi-part upload with a single part. The hashes etc should be the same as far as the client is aware right?

jhford replied:
Well, technically speaking a single part upload is either a PUT to the object *or* a multipart upload with a single part.  Since those are two distinct things under the hood, I'd rather represent them as two different things in the client.  In the final command line tools and libraries that are TC specific, this will all be hidden away.

jonasfj commented on src/client.js:
```diff
@@ -0,0 +1,275 @@
+    let parts = [];
+
+    if (partsize/size < 2) {
+      throw new Error('Multipart upload must have at least 2 parts');
```
if we remove this and allow `partcount == 1` then we don't need the singlepart thing...

jhford replied:
As above, we're then confusing PUTs to the object and multipart uploads with a single part.  I'd agree that single-part upload without using PUT would be weird (enough that I disallow them) but I think were we can precisely match the semantics of the underlying API i'd rather do that.

jonasfj commented on src/client.js:
```diff
@@ -0,0 +1,275 @@
+      // an ETag will just cause the upload to fail for a multipart
+      // upload because it cannot commit the upload and it is not
+      // important for a single part upload
+      etags.push(etag || 'NOETAG');
```
let's just ship it from single part uploads too... then it's consistent...
And let's fallback to empty string if nothing is given...

jhford replied:
it happens for both multipart and singlepart, the only place where the two are distinct is the prepare method.

jonasfj commented on src/client.js:
```diff
@@ -0,0 +1,275 @@
+      await InterchangeFormat.validate(req);
+    }
+
+    if (!parts) {
```
Let's change the format... so in case of single part upload we just provide `parts` as an array of a single element...

Also I have no idea why there is `start`, we can compute that from the previous sizes right?
`_.sumBy(parts.slice(0, i), 'size')` should give you start of part i.

jhford replied:
That implies that all parts are exactly the same, which is not strictly true.  We could have parts where the first is 5mb, then 6mb then 7mb, then 5gb then back to 5mb.  Rather than having to iterate through the list to sum up all the preceeding element sizes, I'd rather be able to iteration and create the filestreams based on the object in the collection.

jonasfj reviewed (approved):
This looks really good :) I think you should implement the queue side, and swing back around and make the client library even more awesome (retries, etc, or maybe I missed that)...

I'm more comfortable r+'ing the interchange format when I see the JSON schema definition in taskcluster-queue. It's a bit more concise and there is a lot of code to navigate here.
